src/txmempool.cpp

```diff
@@ -120,47 +120,91 @@ void CTxMemPool::UpdateForDescendants(txiter updateIt, cacheMap& cachedDescendan
                                       const std::set<uint256>& setExclude, std::set<uint256>& descendants_to_remove,
                                       uint64_t ancestor_size_limit, uint64_t ancestor_count_limit)
 {
-    CTxMemPoolEntry::Children stageEntries, descendants;
-    stageEntries = updateIt->GetMemPoolChildrenConst();
-
-    while (!stageEntries.empty()) {
-        const CTxMemPoolEntry& descendant = *stageEntries.begin();
-        descendants.insert(descendant);
-        stageEntries.erase(descendant);
-        const CTxMemPoolEntry::Children& children = descendant.GetMemPoolChildrenConst();
-        for (const CTxMemPoolEntry& childEntry : children) {
-            cacheMap::iterator cacheIt = cachedDescendants.find(mapTx.iterator_to(childEntry));
-            if (cacheIt != cachedDescendants.end()) {
-                // We've already calculated this one, just add the entries for this set
-                // but don't traverse again.
-                for (txiter cacheEntry : cacheIt->second) {
-                    descendants.insert(*cacheEntry);
+    const CTxMemPoolEntry::Children& children = updateIt->GetMemPoolChildrenConst();
+
+    // initialize to hold all direct children
+    // children guaranteed to be unique at this point
+    std::vector<txiter> descendants(children.size());
+    size_t i = 0;
+    for (const auto& child: children) {
+        descendants[i] = mapTx.iterator_to(child);
+        ++i;
+    }
+
+    {
+        WITH_FRESH_EPOCH(m_epoch);
+        // visit all children
+        for (auto& child : descendants) {
+            visited(child);
+        }
+        for (size_t i = 0, n_to_process = descendants.size(); i < n_to_process; ++i) {
+            const CTxMemPoolEntry& descendant = *descendants[i];
+            const CTxMemPoolEntry::Children& children = descendant.GetMemPoolChildrenConst();
+            for (const CTxMemPoolEntry& childEntry : children) {
+                cacheMap::iterator cacheIt = cachedDescendants.find(mapTx.iterator_to(childEntry));
+                if (cacheIt != cachedDescendants.end()) {
+                    // We've already calculated this one, just add the entries for this set
+                    // but don't traverse again.
+                    for (txiter cacheEntry : cacheIt->second) {
+                        // Add all to descendants which have not yet been added
+                        if (!visited(cacheEntry)) {
+                            descendants.emplace_back(cacheEntry);
+                            // skip self-swap because of buggy std::swap implementations
+                            // on some platforms
+                            if (!(descendants.size() == i+2)) {
+                                std::swap(descendants[i+1], descendants.back());
+                            }
```

Comment:
> we are keeping a "partition" of processed and unprocessed elements.

Yeah that's what I meant by ordering, sorry for being unclear. Thanks for elaborating, this is helpful

Reply:
By preserve ordering, I merely meant that you're keeping the unprocessed elements behind the processed ones. We're saying the same thing, sorry for being unclear.

> If we were to try to do the same with inserts, it would cause N^2 behavior.

Sure, this is the case if you're shifting everything per-insert inside the vector. This is why I suggested using a deque, where it's O(1) per element. It'd be the same performance as what you're doing here, except you just call `insert` instead of swapping. But my concerns with the complexity/readability of the code will be gone if you just comment what you're doing here.

Reply:
yeah deque is same complexity, but it's not the same performance, so it makes sense to continue using a vec even if we have to do a little bit of index management. can add a comment.